compact.go

```diff
@@ -266,20 +266,27 @@ func (c *LeveledCompactor) selectDirs(ds []dirMeta) []dirMeta {
 	return nil
 }
 
-// selectOverlappingDirs returns all dirs with overlaping time ranges.
-// It expects sorted input by mint.
+// selectOverlappingDirs returns all dirs with overlapping time ranges.
+// It expects sorted input by mint and returns the overlapping dirs in the same order as received.
 func (c *LeveledCompactor) selectOverlappingDirs(ds []dirMeta) []string {
 	if len(ds) < 2 {
 		return nil
 	}
 	var overlappingDirs []string
 	globalMaxt := ds[0].meta.MaxTime
+	chained := false
 	for i, d := range ds[1:] {
 		if d.meta.MinTime < globalMaxt {
-			if len(overlappingDirs) == 0 { // When it is the first overlap, need to add the last one as well.
+			if !chained {
 				overlappingDirs = append(overlappingDirs, ds[i].dir)
+				chained = true
 			}
 			overlappingDirs = append(overlappingDirs, d.dir)
+		} else {
+			if chained {
```

Comment:
We can get rid of `chained` variable by having `len(overlappingDirs) > 0` check here instead.